cmd/soroban-cli/src/commands/contract/fetch.rs

```diff
@@ -0,0 +1,235 @@
+    MissingResult,
+    #[error("unexpected contract code data type: {0:?}")]
+    UnexpectedContractCodeDataType(LedgerEntryData),
+    #[error("reading file {0:?}: {1}")]
+    CannotWriteContractFile(PathBuf, io::Error),
+    #[error("cannot parse contract ID {0}: {1}")]
+    CannotParseContractId(String, DecodeError),
+    #[error("network details not provided")]
+    NetworkNotProvided,
+    #[error(transparent)]
+    Network(#[from] network::Error),
+    #[error(transparent)]
+    Ledger(#[from] ledger_file::Error),
+    #[error("cannot create contract directory for {0:?}")]
+    CannotCreateContractDir(PathBuf),
+}
+
+impl From<Infallible> for Error {
+    fn from(_: Infallible) -> Self {
+        unreachable!()
+    }
+}
+
+impl Cmd {
+    pub async fn run(&self) -> Result<(), Error> {
+        let bytes = self.get_bytes().await?;
+        if let Some(out_file) = &self.out_file {
+            if let Some(parent) = out_file.parent() {
+                if !parent.exists() {
+                    fs::create_dir_all(parent)
+                        .map_err(|_| Error::CannotCreateContractDir(out_file.clone()))?;
+                }
+            }
+            fs::write(out_file, bytes)
+                .map_err(|io| Error::CannotWriteContractFile(out_file.clone(), io))
+        } else {
+            let stdout = std::io::stdout();
+            let mut handle = stdout.lock();
+            handle.write_all(&bytes)?;
+            handle.flush()?;
+            Ok(())
+        }
+    }
+
+    pub async fn get_bytes(&self) -> Result<Vec<u8>, Error> {
+        if self.network.is_no_network() {
+            self.run_in_sandbox()
+        } else {
+            self.run_against_rpc_server().await
+        }
+    }
+
+    pub fn network(&self) -> Result<Network, Error> {
+        Ok(self.network.get(&self.locator)?)
+    }
+
+    pub async fn run_against_rpc_server(&self) -> Result<Vec<u8>, Error> {
+        let network = self.network()?;
+        tracing::trace!(?network);
+        let contract_id = self.contract_id()?;
+        let client = Client::new(&network.rpc_url)?;
+        // async closures are not yet stable
+        get_remote_wasm(&client, &contract_id).await
+    }
+
+    pub fn get_state(&self) -> Result<LedgerSnapshot, Error> {
+        Ok(self.ledger_file.read(&self.locator.config_dir()?)?)
+    }
+
+    pub fn run_in_sandbox(&self) -> Result<Vec<u8>, Error> {
+        let contract_id = self.contract_id()?;
+        // Initialize storage and host
+        let snap = Rc::new(self.get_state()?);
+        let mut storage = Storage::with_recording_footprint(snap);
+        Ok(get_contract_wasm_from_storage(&mut storage, contract_id)?)
+    }
+
+    fn contract_id(&self) -> Result<[u8; 32], Error> {
+        utils::contract_id_from_str(&self.contract_id)
+            .map_err(|e| Error::CannotParseContractId(self.contract_id.clone(), e))
+    }
+}
+
+async fn get_remote_wasm(client: &Client, contract_id: &[u8; 32]) -> Result<Vec<u8>, Error> {
```

Comment:
Can we make this a method on the `Client`? Then we could re-use it for `invoke` subcommand where we do: `get_remote_contract_spec_entries` as well.